Tensile/SolutionStructs.py

```diff
@@ -216,37 +244,181 @@ def __init__(self, problemTypeOut, convolutionType, config):
-    problemTypeOut["IndexAssignmentsB"] = [x[0] for x in self.indexB[::-1]]
+      problemTypeOut["NumIndicesC"] = 2+len(filterDims)
 
+    # convert from convolution order to tensor order:
+    self.indexA.reverse()
+    self.indexB.reverse()
+
+    problemTypeOut["IndexAssignmentsA"] = [x[0] for x in self.indexA]
+    problemTypeOut["IndexAssignmentsB"] = [x[0] for x in self.indexB]
+    problemTypeOut["UseBeta"] = False # MI kernels don't use beta
 
     problemTypeOut["SetConstStrideA"]=[]
     for (idx,fbs,dim) in self.indexA:
       if dim.strideA != -1:
         problemTypeOut["SetConstStrideA"].append([idx,dim.strideA])
+    problemTypeOut["SetConstStrideA"].sort()
 
     problemTypeOut["SetConstStrideB"]=[]
     for (idx,fbs,dim) in self.indexB:
       if dim.strideB != -1:
         problemTypeOut["SetConstStrideB"].append([idx,dim.strideB])
+    problemTypeOut["SetConstStrideB"].sort()
 
-    if 0 in [x[0] for x in problemTypeOut["SetConstStrideA"]] or \
-       0 in [x[0] for x in problemTypeOut["SetConstStrideB"]]:
-     print ("warning: need UseInitialStrides")
-     #problemTypeOut["UseInitialStrides"] = 1
-   #else:
-     #problemTypeOut["UseInitialStrides"] = 0
-
+    if [x for x in problemTypeOut["SetConstStrideA"] if x==[0,1]] or \
+       [x for x in problemTypeOut["SetConstStrideB"] if x==[0,1]]:
+      problemTypeOut["UseInitialStrides"] = False
+    else:
+      problemTypeOut["UseInitialStrides"] = True
 
     #self.printUsage(problemTypeOut)
 
+  def dimIdx(self, convolutionChar):
+    return self.convolutionDims[convolutionChar][0]
+
+  def convolutionChar(self, dimIdx):
+    return self.indexAssignments[dimIdx][1].shortChar
+
+  def makeProblem(self, keepTbd, n, c, k, spatialIn=None):
+    """
+    Generate valid problem dims
+
+    TBD values are assumes to be 1 (filter/dilation/stride) or 0(pad) via abs(..) function
+    """
+    numDims = 1 + max(max([x[0] for x in self.indexA]), max([x[0] for x in self.indexB]))
+    sizes = [-1]*numDims
+    astrides = [-1]*numDims
+
+    sizes[self.convolutionDims['N'][0]]=n
+    sizes[self.convolutionDims['C'][0]]=c
+    sizes[self.convolutionDims['K'][0]]=k
+
+    if spatialIn==None:
+      if self.spatial != None:
+        spatialIn = self.spatial
+      else:
+        if keepTbd:
+          spatialIn = [-1]*self.formatNumSpatialDims
+        else:
+          raise RuntimeError ("problemSize must specify spatial parms or set ConvolutionConfig.spatial")
+
+    if len(spatialIn) != self.formatNumSpatialDims:
+      raise RuntimeError ("len(spatialIn=", spatialIn, ") must match formatNumSpatialDims(%d)"%self.formatNumSpatialDims)
+
+    spatialTbd = not all(i!=-1 for i in spatialIn)
+    filterTbd = not all(i!=-1 for i in self.filter)
+    strideTbd = not all(i!=-1 for i in self.stride)
+    dilationTbd = not all(i!=-1 for i in self.dilation)
+    padTbd = not all(i!=-1 for i in self.padStart) or not all(i!=-1 for i in self.padEnd)
```

Comment:
These could be implemented as properties (with `@property`) and tested.

Reply:
can you say more or provide an example? For example, these are function locals rather than class members - not sure how @ property applies?

Reply:
The 4 that I highlighted are really properties of the object - `filterTbd` is a statement on the value of `self.filter`.  I'm suggesting that these be hoisted outside of this function and made into properties of the class.

    @property
    def filterTbd(self):
        return -1 in self.filter

    # ...
        if keepTbd and self.filterTbd:

Also, when you force-push, it invalidates the links from these comments to the code and makes it difficult to see what's changed.